doc/secure_upgrade/secure_upgrade.md

```diff
@@ -0,0 +1,179 @@
+<!-- /vscode-markdown-toc -->
+
+###  1.1. <a name='Revision'></a>Revision
+  Rev  |  Date   |      Author      | Change Description |
+| :---: | :-----: | :--------------: | ------------------ |
+|  0.1  | 6/2022 |   Yona Cohen   | Phase 1 Design     |
+
+###  1.2. <a name='Scope'></a>Scope
+
+This secure upgrade hld doc described the requirements, architecture, and general flow details of secure upgrade feature in switches SONIC OS based.
+
+###  1.3. <a name='DefinitionsAbbreviations'></a>Definitions/Abbreviations
+
+	SU - secure upgrade
+	SB - secure boot	
+	CMS - Cryptographic Message Syntax
+
+###  1.4. <a name='Overview'></a>Overview
+
+Secure installation and upgrade of OS are setting a requirement to ensure the image was not modified since it was built by the vendor. This is a basic requirement as part of general OS security. In order to help the credibility of the OS image, secure upgrade mechanism guarantees the image was not modified by any unauthorized entity since its original creation.
+
+###  1.5. <a name='Requirements'></a>Requirements
+
+We want to enable secure upgrade of SONIC. This will include secure upgrade of SONIC while installing SONIC via ONIE or SONIC. The feature requires a signing process which will include a private key and certificate. The verification process on the other end requires a public key certificate for validation. Both processes are required to be supported by various scenarios such as development and release.
+In order to do secure installation of SONIC from ONIE, it has to include the flowing secure upgrade verification commit TBD (on pull request - waiting for it to be created)
+Device arch should support secure boot including UEFI tools.
+
+###  1.6. <a name='ArchitectureDesign'></a>Architecture Design
+####  1.6.1. <a name='InstallationFlow'></a>Installation Flow
+![su_installation_flows](su_installation_flows.png)
+
+####  1.6.2. <a name='signing'></a>signing
+![signing_flow](signing_flow.png)
+(flow description in the chapter below)
+####  1.6.3. <a name='verification'></a>verification
+Verification flow calls a verification script during image installation, in (1.7.3.) verification process below.
+
+###  1.7. <a name='High-LevelDesign'></a>High-Level Design
+
+
+This feature has 2 flows to be supported:
+	
+####  1.7.1. <a name='Signingprocess'></a>Signing process  
+
+Sign sonic image during build process in build_image.sh script, right after generation.
+We want to add the flag SECURE_UPGRADE_MODE to allow several modes of secure upgrade. When SECURE_UPGRADE_MODE == 'no_sign' no change will be made to the current build process. In case of SECURE_UPGRADE_MODE == 'dev', image will be signed by development script and when SECURE_UPGRADE_MODE == 'prod' - image will be signed by production script. In addition, we want to add the flags SECURE_UPGRADE_DEV_SIGNING_KEY and SECURE_UPGRADE_DEV_SIGNING_CERT, to provide the option to give path for development keys during the build.
```

Comment:
How is the signing prod process is different from dev? How do you secure production keys? do you use vault kind of infra?

Reply:
The feature is including the code that supports secure upgrade using development keys. And any Vendor should decide his own design flow for production. 
In general, in the build flow when the user is building the image with "prod" flag (as explained in chapter 1.7.1. Signing process) the build flow will call a signing script that should be coded by the vendor, this script should send the image to a different server "sign server" and this server should return the image signed.

Reply:
@davidpil2002 , we have a different environment for build production and dev, production is always built in security environment. Is it possible to use the same process for both production and dev? The only different is the certificates used, but the build steps are all the same.
It is good to provide an additional script to sign images from the "sign server", but can we support to the simple scenario to support "sign server" and build on the same machine? How about removing the "_DEV_" in the parameters?

Reply:
@xumia thank you for the question. In many cases, different customers have different production servers with different environments. Therefor no single script will be applicable here. In the case you suggest - it is possible to copy the dev key as an argument to prod script, but this is a specific use case